src/persist-client/src/machine.rs

```diff
@@ -0,0 +1,400 @@
+                let cap = ReadCapability {
+                    seqno: meta.seqno,
+                    since: Antichain::from(since),
+                };
+                (id.clone(), cap)
+            })
+            .collect();
+        let trace = state
+            .trace
+            .iter()
+            .map(|(key, desc)| {
+                let lower = desc.lower.iter().map(|x| T::decode(*x)).collect::<Vec<_>>();
+                let upper = desc.upper.iter().map(|x| T::decode(*x)).collect::<Vec<_>>();
+                let since = desc.since.iter().map(|x| T::decode(*x)).collect::<Vec<_>>();
+                let desc = Description::new(
+                    Antichain::from(lower),
+                    Antichain::from(upper),
+                    Antichain::from(since),
+                );
+                (key.clone(), desc)
+            })
+            .collect();
+        State {
+            writers,
+            readers,
+            trace,
+        }
+    }
+
+    pub fn to_meta(&self) -> CollectionMeta {
+        CollectionMeta {
+            writers: self
+                .writers
+                .iter()
+                .map(|(id, cap)| {
+                    let upper = cap.upper.elements().iter().map(|x| T::encode(x)).collect();
+                    let meta = WriterMeta {
+                        seqno: cap.seqno,
+                        upper,
+                    };
+                    (id.clone(), meta)
+                })
+                .collect(),
+            readers: self
+                .readers
+                .iter()
+                .map(|(id, cap)| {
+                    let since = cap.since.elements().iter().map(|x| T::encode(x)).collect();
+                    let meta = ReaderMeta {
+                        seqno: cap.seqno,
+                        since,
+                    };
+                    (id.clone(), meta)
+                })
+                .collect(),
+            trace: self
+                .trace
+                .iter()
+                .map(|(key, desc)| {
+                    let desc = TraceBatchMeta {
+                        lower: desc.lower().iter().map(|x| T::encode(x)).collect(),
+                        upper: desc.upper().iter().map(|x| T::encode(x)).collect(),
+                        since: desc.since().iter().map(|x| T::encode(x)).collect(),
+                    };
+                    (key.clone(), desc)
+                })
+                .collect(),
+            key_codec: "WIP".into(),
+            val_codec: "WIP".into(),
+            ts_codec: "WIP".into(),
+            diff_codec: "WIP".into(),
+        }
+    }
+
+    pub fn since(&self) -> Antichain<T> {
+        //  WIP double check this initial and the meet_assign
+        //
+        // WIP what happens if all readers go away and then some come back?
+        let mut ret = Antichain::from_elem(T::minimum());
+        for (_, cap) in self.readers.iter() {
+            ret.meet_assign(&cap.since);
```

Comment:
pretty sure the meet_assign is indeed not going to work here with the given initial value because its like a generalized min - so for 1d times it will stay at [0] as written above

it seems like for the second part - what happens if all the readers leave? i think for that case we'll need to store a collection level since somewhere (ideally near where the finalized trace batches live), so that it can record "the since of this collection has been advanced to <foo>, and no new readers can start up with a since less_than <foo>"? We can also advance that collection level since periodically when there are no readers?

Reply:
> We can also advance that collection level since periodically when there are no readers?
hmm this is a bit sus to me though. my preference is to only do that sort of thing in response to explicit system input, if we can swing it

Reply:
okay maintaining an explicit since now